src/Host/Client/Impl/Host/LocalRHostConnector.cs

```diff
@@ -0,0 +1,174 @@
+
+                try {
+                    server.Start();
+                    break;
+                } catch (SocketException ex) {
+                    if (ex.SocketErrorCode == SocketError.AddressAlreadyInUse) {
+                        server = null;
+                    } else {
+                        throw new MessageTransportException(ex);
+                    }
+                } catch (WebSocketException ex) {
+                    throw new MessageTransportException(ex);
+                }
+            }
+
+            if (server == null) {
+                throw new MessageTransportException(new SocketException((int)SocketError.AddressAlreadyInUse));
+            }
+
+            var psi = new ProcessStartInfo {
+                FileName = rhostExe,
+                UseShellExecute = false
+            };
+
+            var shortHome = new StringBuilder(NativeMethods.MAX_PATH);
+            NativeMethods.GetShortPathName(_rHome, shortHome, shortHome.Capacity);
+            psi.EnvironmentVariables["R_HOME"] = shortHome.ToString();
+
+            psi.EnvironmentVariables["PATH"] = rBinPath + ";" + Environment.GetEnvironmentVariable("PATH");
+
+            if (name != null) {
+                psi.Arguments += " --rhost-name " + name;
+            }
+
+            psi.Arguments += Invariant($" --rhost-connect ws://127.0.0.1:{server.Port}");
+
+            if (!ShowConsole) {
+                psi.CreateNoWindow = true;
+            }
+
+            if (!string.IsNullOrWhiteSpace(rCommandLineArguments)) {
+                psi.Arguments += Invariant($" {rCommandLineArguments}");
+            }
+
+            var process = Process.Start(psi);
+            _log.RHostProcessStarted(psi);
+            process.EnableRaisingEvents = true;
+
+            // Timeout increased to allow more time in test and code coverage runs.
+            await Task.WhenAny(transportTcs.Task, Task.Delay(timeout, cancellationToken)).Unwrap();
+            if (!transportTcs.Task.IsCompleted) {
+                _log.FailedToConnectToRHost();
+                throw new RHostTimeoutException("Timed out waiting for R host process to connect");
+            }
+
+            var cts = new CancellationTokenSource();
+            cts.Token.Register(() => {
+                if (!process.HasExited) {
+                    try {
+                        process.WaitForExit(500);
+                        if (!process.HasExited) {
+                            process.Kill();
+                            process.WaitForExit();
+                        }
+                    } catch (InvalidOperationException) { }
+                }
+                _log.RHostProcessExited();
+            });
+
+            var host = new RHost(name, callbacks, transportTcs.Task.Result, process.Id, cts);
+            process.Exited += delegate { host.Dispose(); };
+            return host;
+
+            //using (this)
```

Comment:
Sorry, garbage :)

---
In reply to: [73022424](https://github.com/Microsoft/RTVS/pull/2133#discussion_r73022424) [](ancestors = 73022424)